File: typescript/example/example_keys.ts
import { ASimpleVectorsClient } from "../asimplevectors/src";

async function testKeyValue(client: ASimpleVectorsClient): Promise<void> {
    const spaceName = "spacename";
    const keyName = "test_key";
    const keyValueData = "Hello, ASimpleVectors!";

    // Step 1: Create a key-value pair in the space
    console.log(`Creating key-value pair in '${spaceName}'`);
    await client.putKeyValue(spaceName, keyName, keyValueData);
    console.log(`Key '${keyName}' created successfully.`);

    // Step 2: Retrieve the key-value pair
    console.log(`Retrieving key-value pair for key '${keyName}'`);
    try {
        const response = await client.getKeyValue(spaceName, keyName);
        if (response) {
            console.log(`Key '${keyName}' retrieved successfully:`, response);
        }
    } catch (error) {
        console.error(`Failed to retrieve key '${keyName}':`, error);
    }

    // Step 3: List all keys in the space
    console.log(`Listing all keys in '${spaceName}'`);
    try {
        const response = await client.listKeys(spaceName);
        if (response && response.keys) {
            console.log(`Keys in '${spaceName}':`, response.keys);
        }
    } catch (error) {
        console.error(`Failed to list keys in '${spaceName}':`, error);
    }

    // Step 4: Delete the key-value pair
    console.log(`Deleting key '${keyName}' from '${spaceName}'`);
    try {
        await client.deleteKeyValue(spaceName, keyName);
        console.log(`Key '${keyName}' deleted successfully.`);
    } catch (error) {
        console.error(`Failed to delete key '${keyName}':`, error);
    }

    // Step 5: Verify the key is deleted
    console.log(`Verifying deletion of key '${keyName}'`);
    try {
        // Attempt to delete the key again to confirm its absence
        await client.deleteKeyValue(spaceName, keyName);
        console.error(`Unexpected behavior: Key '${keyName}' was deleted twice.`);
    } catch (error) {
        if (error.response?.status === 400) {
            // Expected behavior: Key does not exist
            console.log(`Key '${keyName}' successfully deleted and verified.`);
        } else {
            // Unexpected error
            console.error(`Failed to verify deletion of key '${keyName}':`, error);
        }
    }
}

async function main(): Promise<void> {
    const client = new ASimpleVectorsClient("localhost", 21001);
    try {
        // Run the key-value storage test
        await testKeyValue(client);
    } finally {
        console.log("Closing the client.");
        // Ensure any necessary cleanup is handled here
    }
}

// Run the example
main().catch((error) => {
    console.error("An error occurred:", error);
});

File: python/example/example_keys.py
import asyncio
from asimplevectors.client import ASimpleVectorsClient, KeyNotFoundError

# Test function to create, read, list, and delete key-value pairs
async def test_keyvalue(client: ASimpleVectorsClient):
    # Step 1: Create a key-value pair in the space
    print("Creating key-value pair in 'spacename'")
    key_name = "test_key"
    key_value_data = {
        "text": "Hello, ASimpleVectors!"
    }
    await client.put_key_value("spacename", key_name, key_value_data)
    print(f"Key '{key_name}' created successfully.")

    # Step 2: Retrieve the key-value pair
    print(f"Retrieving key-value pair for key '{key_name}'")
    response = await client.get_key_value("spacename", key_name)
    if response:
        print(f"Key '{key_name}' retrieved successfully:", response)

    # Step 3: List all keys in the space
    print("Listing all keys in 'spacename'")
    response = await client.list_keys("spacename")
    if response:
        print("Keys in 'spacename':", response.keys)

    # Step 4: Delete the key-value pair
    print(f"Deleting key '{key_name}' from 'spacename'")
    await client.delete_key_value("spacename", key_name)
    print(f"Key '{key_name}' deleted successfully.")

    # Step 5: Verify the key is deleted
    print(f"Verifying deletion of key '{key_name}'")
    try:
        response = await client.get_key_value("spacename", key_name)
        print(f"Key '{key_name}' still exists:", response)
    except KeyNotFoundError:
        print(f"Key '{key_name}' successfully deleted.")

async def main():
    client = ASimpleVectorsClient(host="localhost")
    try:
        # Run the key-value storage test
        await test_keyvalue(client)
    finally:
        # Ensure the client session is closed
        await client.close()

if __name__ == "__main__":
    asyncio.run(main())
